File: sdks/go/skopli/smoke_test.go
// Smoke test driving the Go facade over the C ABI end to end: read -> rollup ->
// price against the committed golden case, asserting structural parity with the
// shared conformance gold (the same golden/claude/basic + pricing/catalogs data
// the Rust capi smoke test uses). Also verifies the errors.Is sentinel mapping
// for an invalid-argument and a catalog error.
//
// The test needs the GNU-target cdylib on the loader path at run time; see the
// package README for the CGO/PATH mechanics.
package skopli

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"reflect"
	"runtime"
	"sort"
	"testing"
)

// repoRoot walks up from this test file to the workspace root (four parents:
// smoke_test.go -> skopli -> go -> sdks -> <root>).
func repoRoot(t *testing.T) string {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("cannot resolve caller file")
	}
	// file = <root>/sdks/go/skopli/smoke_test.go
	return filepath.Clean(filepath.Join(filepath.Dir(file), "..", "..", ".."))
}

func claudeInputDir(t *testing.T) string {
	return filepath.Join(repoRoot(t), "golden", "claude", "basic", "input")
}

func claudeReadOptions(t *testing.T) ReadUsageOptions {
	return ReadUsageOptions{
		Options: Options{
			Home: "/nonexistent",
			Env:  map[string]string{"CLAUDE_CONFIG_DIR": claudeInputDir(t)},
		},
		Harnesses: []Harness{HarnessClaude},
		TZ:        "UTC",
	}
}

// --- meta -------------------------------------------------------------------

func TestMeta(t *testing.T) {
	if got := ABIVersion(); got != 1 {
		t.Fatalf("ABIVersion = %d, want 1", got)
	}
	if got := SchemaVersion(); got != 1 {
		t.Fatalf("SchemaVersion = %d, want 1", got)
	}
	if Version() == "" {
		t.Fatal("Version is empty")
	}
	dir, err := DefaultCacheDir()
	if err != nil {
		t.Fatalf("DefaultCacheDir: %v", err)
	}
	if dir == "" {
		t.Fatal("DefaultCacheDir is empty")
	}
}

// --- harness enum parity ----------------------------------------------------

// TestHarnessEnumCoversRegistry asserts KnownHarnesses covers every id in the
// shared registry-id fixture, so a reader added to the core cannot silently
// fall out of the Go surface.
func TestHarnessEnumCoversRegistry(t *testing.T) {
	path := filepath.Join(repoRoot(t), "golden", "registry", "ids.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read registry ids fixture: %v", err)
	}
	var fixture struct {
		IDs []string `json:"ids"`
	}
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatalf("parse registry ids fixture: %v", err)
	}
	known := make(map[Harness]bool, len(KnownHarnesses))
	for _, h := range KnownHarnesses {
		known[h] = true
	}
	for _, id := range fixture.IDs {
		if !known[Harness(id)] {
			t.Errorf("KnownHarnesses missing registry id %q", id)
		}
	}
	registered := make(map[string]bool, len(fixture.IDs))
	for _, id := range fixture.IDs {
		registered[id] = true
	}
	for _, h := range KnownHarnesses {
		if !registered[string(h)] {
			t.Errorf("KnownHarnesses has id %q the registry does not", h)
		}
	}
}

// --- detect -----------------------------------------------------------------

func TestDetectFindsClaude(t *testing.T) {
	opts := claudeReadOptions(t).Options
	detection, err := DetectHarnesses(&opts)
	if err != nil {
		t.Fatalf("DetectHarnesses: %v", err)
	}
	found := false
	for _, h := range detection.Supported {
		if h == HarnessClaude {
			found = true
		}
	}
	if !found {
		t.Fatalf("claude not in supported: %+v", detection)
	}
	if len(detection.Unsupported) != 0 {
		t.Fatalf("unsupported should be empty over this ABI: %+v", detection.Unsupported)
	}
}

// --- cost_usd ---------------------------------------------------------------

func TestCostUSDFlat(t *testing.T) {
	tokens := TokenCounts{Input: 1000, Output: 500}
	price := ModelPrice{Input: 1.25, Output: 10.0}
	got, err := CostUSD(tokens, price)
	if err != nil {
		t.Fatalf("CostUSD: %v", err)
	}
	want := (1000.0*1.25 + 500.0*10.0) / 1_000_000.0
	if got != want {
		t.Fatalf("CostUSD = %v, want %v", got, want)
	}
}

// --- read -> rollup round trip vs gold --------------------------------------

func TestReadRollupRoundtripMatchesGold(t *testing.T) {
	result, err := ReadUsage(claudeReadOptions(t))
	if err != nil {
		t.Fatalf("ReadUsage: %v", err)
	}
	if len(result.Events) == 0 {
		t.Fatal("read produced no events")
	}

	// Sort events the way the exporter/gold does before rolling up.
	events := append([]UsageEvent(nil), result.Events...)
	sort.SliceStable(events, func(i, j int) bool {
		a, b := events[i], events[j]
		if a.Timestamp != b.Timestamp {
			return a.Timestamp < b.Timestamp
		}
		if a.SessionID != b.SessionID {
			return a.SessionID < b.SessionID
		}
		if a.MessageID != b.MessageID {
			return a.MessageID < b.MessageID
		}
		return a.Model < b.Model
	})

	byModel, err := Rollup(events, RollupOptions{By: ByModel, TZ: "UTC"})
	if err != nil {
		t.Fatalf("Rollup: %v", err)
	}

	// Load the gold model lane and compare structurally (round-trip both sides
	// through the wire so pointer/omitempty encoding differences vanish).
	goldPath := filepath.Join(repoRoot(t), "golden", "claude", "basic", "expected-rollup.json")
	goldBytes, err := os.ReadFile(goldPath)
	if err != nil {
		t.Fatalf("read gold: %v", err)
	}
	var gold struct {
		By struct {
			Model []RollupBucket `json:"model"`
		} `json:"by"`
	}
	if err := json.Unmarshal(goldBytes, &gold); err != nil {
		t.Fatalf("decode gold: %v", err)
	}

	if !reflect.DeepEqual(normalize(t, byModel), normalize(t, gold.By.Model)) {
		t.Fatalf("rollup(by=model) mismatch:\n got: %s\nwant: %s",
			mustJSON(t, byModel), mustJSON(t, gold.By.Model))
	}
}

// --- pricing round trip -----------------------------------------------------

func readCatalogPayload(t *testing.T, name string) json.RawMessage {
	t.Helper()
	p := filepath.Join(repoRoot(t), "golden", "pricing", "catalogs", name+".json")
	b, err := os.ReadFile(p)
	if err != nil {
		t.Fatalf("read catalog %s: %v", name, err)
	}
	return json.RawMessage(b)
}

// pricingEvents mirrors the synthetic events from the pricing conformance case.
func pricingEvents() []UsageEvent {
	ev := func(mid, ts, model string, tk TokenCounts) UsageEvent {
		return UsageEvent{
			Harness:   HarnessOpencode,
			Timestamp: ts,
			SessionID: "s",
			MessageID: mid,
			Turn:      true,
			Model:     model,
			Tokens:    tk,
		}
	}
	u := func(v uint64) *uint64 { return &v }
	return []UsageEvent{
		ev("flat", "2026-08-01T00:00:00.000Z", "gpt-5",
			TokenCounts{Input: 1000, Output: 500}),
		ev("tiered", "2026-08-01T00:01:00.000Z", "claude-sonnet-4-5",
			TokenCounts{Input: 200000, Output: 10000, CacheRead: 20000, CacheWrite: 30000, CacheWrite1h: u(10000), Reasoning: 2000}),
		ev("base-1h", "2026-08-01T00:02:00.000Z", "claude-sonnet-4-5",
			TokenCounts{Input: 5000, Output: 1000, CacheRead: 2000, CacheWrite: 4000, CacheWrite1h: u(1500)}),
		ev("alias", "2026-08-01T00:03:00.000Z", "us.anthropic.claude-opus-4-6-20260115-v1:0",
			TokenCounts{Input: 800, Output: 200}),
		ev("miss", "2026-08-01T00:04:00.000Z", "totally-unknown-model-9000",
			TokenCounts{Input: 100, Output: 100}),
	}
}

func TestPricingRoundtrip(t *testing.T) {
	const pinned = "2026-08-01T00:00:00.000Z"
	pricing, err := NewPricing(PricingOptions{
		Mode:    ModeCalculate,
		Sources: []PricingSource{},
		Catalogs: []Catalog{
			{Source: "openrouter", FetchedAt: pinned, Format: "openrouter", Payload: readCatalogPayload(t, "openrouter")},
			{Source: "litellm", FetchedAt: pinned, Format: "litellm", Payload: readCatalogPayload(t, "litellm")},
		},
	})
	if err != nil {
		t.Fatalf("NewPricing: %v", err)
	}
	defer pricing.Close()

	infos, err := pricing.Catalogs()
	if err != nil {
		t.Fatalf("Catalogs: %v", err)
	}
	if len(infos) != 2 {
		t.Fatalf("want 2 catalogs, got %d: %+v", len(infos), infos)
	}
	if infos[0].Source != "openrouter" || infos[1].Source != "litellm" {
		t.Fatalf("catalog sources mismatch: %+v", infos)
	}
	if infos[0].Models == 0 {
		t.Fatalf("openrouter catalog has no models: %+v", infos[0])
	}

	priced, err := pricing.PriceEvents(pricingEvents(), RollupOptions{By: ByModel})
	if err != nil {
		t.Fatalf("PriceEvents: %v", err)
	}
	if len(priced) == 0 {
		t.Fatal("no priced groups")
	}

	var gpt5, miss *PricedRollup
	for i := range priced {
		switch priced[i].Key {
		case "gpt-5":
			gpt5 = &priced[i]
		case "totally-unknown-model-9000":
			miss = &priced[i]
		}
	}
	if gpt5 == nil {
		t.Fatal("gpt-5 group missing")
	}
	if !gpt5.Pricing.Priced {
		t.Fatalf("gpt-5 should be priced: %+v", gpt5.Pricing)
	}
	if gpt5.Pricing.USD == nil || *gpt5.Pricing.USD <= 0 {
		t.Fatalf("gpt-5 usd should be > 0: %+v", gpt5.Pricing)
	}
	if miss == nil {
		t.Fatal("miss group missing")
	}
	if miss.Pricing.Priced {
		t.Fatalf("unknown model should be a miss: %+v", miss.Pricing)
	}
}

// --- error sentinel mapping -------------------------------------------------

func TestErrInvalidArgument(t *testing.T) {
	_, err := ReadUsage(ReadUsageOptions{Since: "not-a-date"})
	if err == nil {
		t.Fatal("expected an error for a bad since date")
	}
	if !errors.Is(err, ErrInvalidArgument) {
		t.Fatalf("want ErrInvalidArgument, got %v", err)
	}
}

func TestErrCatalog(t *testing.T) {
	// A catalog with an unknown format triggers AG_STATUS_CATALOG.
	_, err := NewPricing(PricingOptions{
		Sources: []PricingSource{},
		Catalogs: []Catalog{
			{Source: "bogus", Format: "not-a-real-format", Payload: json.RawMessage(`{}`)},
		},
	})
	if err == nil {
		t.Fatal("expected a catalog error")
	}
	if !errors.Is(err, ErrCatalog) {
		t.Fatalf("want ErrCatalog, got %v", err)
	}
}

// --- helpers ----------------------------------------------------------------

// normalize round-trips a value through JSON so pointer vs value and omitempty
// encoding differences do not defeat DeepEqual.
func normalize(t *testing.T, v any) any {
	t.Helper()
	b := mustJSON(t, v)
	var out any
	if err := json.Unmarshal(b, &out); err != nil {
		t.Fatalf("normalize decode: %v", err)
	}
	return out
}

func mustJSON(t *testing.T, v any) []byte {
	t.Helper()
	b, err := json.Marshal(v)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	return b
}

File: sdks/csharp/Skopli.Tests/SmokeTest.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Skopli;
using Xunit;

namespace Skopli.Tests;

/// <summary>
/// Drives the C# facade end to end through the csbindgen P/Invoke layer against
/// the committed conformance gold - the same data the Rust capi smoke test
/// and the Go/Java SDKs exercise. Asserts structural parity
/// of the read -> rollup -> price pipeline versus the golden expected JSON.
/// </summary>
public sealed class SmokeTest
{
    private static string RepoRoot()
    {
        // Walk up from the test assembly location until the golden/ dir appears.
        string dir = AppContext.BaseDirectory;
        while (dir is not null && !Directory.Exists(Path.Combine(dir, "golden")))
        {
            dir = Path.GetDirectoryName(dir)!;
        }
        Assert.False(dir is null, "could not locate repo root (golden/)");
        return dir!;
    }

    private static string GoldenPath(params string[] parts) =>
        Path.Combine(new[] { RepoRoot(), "golden" }.Concat(parts).ToArray());

    // -- harness enum parity ------------------------------------------------

    [Fact]
    public void HarnessEnumCoversRegistry()
    {
        // The shared registry-id fixture is the canonical list of harness ids
        // the core registers; the Harness enum must cover every one (the enum
        // names lowercase to the wire ids) so a reader added to the core cannot
        // silently fall out of the typed surface.
        string json = File.ReadAllText(GoldenPath("registry", "ids.json"));
        string[] ids = JsonNode.Parse(json)!["ids"]!.AsArray()
            .Select(n => n!.GetValue<string>())
            .ToArray();
        var known = Enum.GetNames<Harness>()
            .Select(name => name.ToLowerInvariant())
            .Where(name => name != "other")
            .ToHashSet();
        string[] missing = ids.Where(id => !known.Contains(id)).ToArray();
        Assert.True(missing.Length == 0, $"Harness enum missing registry ids: {string.Join(", ", missing)}");
        string[] extra = known.Except(ids).OrderBy(id => id).ToArray();
        Assert.True(extra.Length == 0, $"Harness enum has ids the registry does not: {string.Join(", ", extra)}");
    }

    // -- meta / lifecycle ---------------------------------------------------

    [Fact]
    public void Meta()
    {
        Assert.Equal(1u, SkopliClient.AbiVersion);
        Assert.Equal(1u, SkopliClient.SchemaVersion);
        Assert.False(string.IsNullOrEmpty(SkopliClient.Version));
        Assert.Contains("skopli", SkopliClient.DefaultCacheDir());
    }

    // -- cost_usd -----------------------------------------------------------

    [Fact]
    public void CostUsdFlat()
    {
        double usd = SkopliClient.CostUsd(
            new TokenCounts { Input = 1000, Output = 500 },
            new ModelPrice { Input = 1.25, Output = 10.0 });
        double expected = (1000.0 * 1.25 + 500.0 * 10.0) / 1_000_000.0;
        Assert.Equal(expected, usd, 12);
    }

    // -- detect / read / rollup over golden/claude/basic --------------------

    private static ReadUsageOptions ClaudeOptions()
    {
        string inputDir = GoldenPath("claude", "basic", "input");
        return new ReadUsageOptions
        {
            Home = "/nonexistent",
            Env = new Dictionary<string, string> { ["CLAUDE_CONFIG_DIR"] = inputDir },
            Harnesses = new[] { "claude" },
            Tz = "UTC",
        };
    }

    [Fact]
    public void DetectFindsClaude()
    {
        string inputDir = GoldenPath("claude", "basic", "input");
        Detection detection = SkopliClient.DetectHarnesses(new DetectOptions
        {
            Home = "/nonexistent",
            Env = new Dictionary<string, string> { ["CLAUDE_CONFIG_DIR"] = inputDir },
        });
        Assert.Contains("claude", detection.Supported);
        Assert.Empty(detection.Unsupported);
    }

    [Fact]
    public void ReadRollupRoundtripMatchesGold()
    {
        ReadUsageResult envelope = SkopliClient.ReadUsage(ClaudeOptions());
        Assert.NotEmpty(envelope.Events);

        // Sort events the way the exporter/gold does before rolling up.
        List<UsageEvent> sorted = envelope.Events
            .OrderBy(e => e.Timestamp, StringComparer.Ordinal)
            .ThenBy(e => e.SessionId, StringComparer.Ordinal)
            .ThenBy(e => e.MessageId, StringComparer.Ordinal)
            .ThenBy(e => e.Model, StringComparer.Ordinal)
            .ToList();

        IReadOnlyList<Rollup> byModel = SkopliClient.Rollup(sorted, new RollupOptions { By = RollupBy.Model });

        // Compare (JSON-normalized) to the committed gold's model lane.
        JsonNode gold = JsonNode.Parse(File.ReadAllText(GoldenPath("claude", "basic", "expected-rollup.json")))!;
        JsonNode goldModel = gold["by"]!["model"]!;

        Assert.Equal(Normalize(goldModel), NormalizeList(byModel));
    }

    // -- pricing round-trip over golden/pricing/catalogs --------------------

    private static string ReadCatalog(string name) =>
        File.ReadAllText(GoldenPath("pricing", "catalogs", $"{name}.json"));

    private static IReadOnlyList<UsageEvent> PricingEvents()
    {
        UsageEvent Ev(string mid, int sec, string model, TokenCounts tokens) => new()
        {
            Harness = "opencode",
            Timestamp = $"2026-08-01T00:0{sec}:00.000Z",
            SessionId = "s",
            MessageId = mid,
            Turn = true,
            Subagent = false,
            Model = model,
            Tokens = tokens,
        };

        return new[]
        {
            Ev("flat", 0, "gpt-5", new TokenCounts { Input = 1000, Output = 500 }),
            Ev("tiered", 1, "claude-sonnet-4-5", new TokenCounts
            {
                Input = 200_000, Output = 10_000, CacheRead = 20_000,
                CacheWrite = 30_000, CacheWrite1h = 10_000, Reasoning = 2_000,
            }),
            Ev("base-1h", 2, "claude-sonnet-4-5", new TokenCounts
            {
                Input = 5_000, Output = 1_000, CacheRead = 2_000,
                CacheWrite = 4_000, CacheWrite1h = 1_500,
            }),
            Ev("alias", 3, "us.anthropic.claude-opus-4-6-20260115-v1:0", new TokenCounts { Input = 800, Output = 200 }),
            Ev("miss", 4, "totally-unknown-model-9000", new TokenCounts { Input = 100, Output = 100 }),
        };
    }

    [Fact]
    public void PricingRoundtrip()
    {
        const string pinned = "2026-08-01T00:00:00.000Z";
        var options = new PricingOptions
        {
            Mode = PricingMode.Calculate,
            Sources = Array.Empty<PricingSource>(),
            Catalogs = new[]
            {
                new Catalog { Source = "openrouter", FetchedAt = pinned, Format = "openrouter", Payload = ReadCatalog("openrouter") },
                new Catalog { Source = "litellm", FetchedAt = pinned, Format = "litellm", Payload = ReadCatalog("litellm") },
            },
        };

        using Pricing pricing = SkopliClient.CreatePricing(options);

        IReadOnlyList<CatalogInfo> infos = pricing.Catalogs();
        Assert.Equal(2, infos.Count);
        Assert.Equal("openrouter", infos[0].Source);
        Assert.Equal("litellm", infos[1].Source);
        Assert.True(infos[0].Models > 0);

        IReadOnlyList<PricedRollup> priced = pricing.PriceEvents(PricingEvents(), new RollupOptions { By = RollupBy.Model });
        Assert.NotEmpty(priced);

        // gpt-5 is a flat priced hit; the unknown model is a miss.
        PricedRollup gpt5 = priced.Single(g => g.Rollup.Key == "gpt-5");
        PriceHit gpt5Hit = Assert.IsType<PriceHit>(gpt5.Pricing);
        Assert.True(gpt5Hit.Priced);
        Assert.Equal(0.00625, gpt5Hit.Usd, 9);

        PricedRollup miss = priced.Single(g => g.Rollup.Key == "totally-unknown-model-9000");
        PriceMiss missLookup = Assert.IsType<PriceMiss>(miss.Pricing);
        Assert.False(missLookup.Priced);

        // Structural parity vs the committed gold's exactly-agreeing groups
        // (the flat gpt-5 hit, the aliased Bedrock->anthropic key, and the miss).
        // The tiered-aggregate group differs by construction between price_events
        // and the gold's price_rollups, so it is
        // spot-checked rather than full-compared.
        JsonNode gold = JsonNode.Parse(File.ReadAllText(GoldenPath("pricing", "basic", "expected-priced-rollup.json")))!;
        JsonArray goldRollups = gold["rollups"]!.AsArray();

        JsonObject goldAlias = goldRollups.Single(n => (string?)n!["key"] == "us.anthropic.claude-opus-4-6-20260115-v1:0")!.AsObject();
        PricedRollup alias = priced.Single(g => g.Rollup.Key == "us.anthropic.claude-opus-4-6-20260115-v1:0");
        PriceHit aliasHit = Assert.IsType<PriceHit>(alias.Pricing);
        Assert.Equal("anthropic/claude-opus-4.6", aliasHit.Key);
        Assert.Equal((double)goldAlias["pricing"]!["usd"]!, aliasHit.Usd, 9);
        Assert.Equal((string?)goldAlias["pricing"]!["source"], aliasHit.Source);

        // Key-set + count parity across all groups vs the gold.
        var goldKeys = goldRollups.Select(n => (string)n!["key"]!).OrderBy(k => k, StringComparer.Ordinal).ToList();
        var gotKeys = priced.Select(g => g.Rollup.Key).OrderBy(k => k, StringComparer.Ordinal).ToList();
        Assert.Equal(goldKeys, gotKeys);
    }

    [Fact]
    public void PricingDisposesIdempotently()
    {
        Pricing pricing = SkopliClient.CreatePricing(new PricingOptions
        {
            Mode = PricingMode.Calculate,
            Sources = Array.Empty<PricingSource>(),
            Catalogs = new[]
            {
                new Catalog { Source = "litellm", FetchedAt = "2026-08-01T00:00:00.000Z", Format = "litellm", Payload = ReadCatalog("litellm") },
            },
        });
        pricing.Dispose();
        pricing.Dispose(); // idempotent
        Assert.Throws<ObjectDisposedException>(() => pricing.Catalogs());
    }

    // -- forced errors ------------------------------------------------------

    [Fact]
    public void InvalidSinceThrowsInvalidArgument()
    {
        var options = ClaudeOptions() with { Since = "not-a-date" };
        Assert.Throws<InvalidArgumentException>(() => SkopliClient.ReadUsage(options));
    }

    [Fact]
    public void UnknownCatalogFormatThrowsCatalog()
    {
        var options = new PricingOptions
        {
            Mode = PricingMode.Calculate,
            Sources = Array.Empty<PricingSource>(),
            Catalogs = new[]
            {
                new Catalog { Source = "bogus", Format = "nonsense-format", Payload = "{}" },
            },
        };
        Assert.Throws<CatalogException>(() => SkopliClient.CreatePricing(options));
    }

    // -- normalization helpers ----------------------------------------------

    private static readonly JsonSerializerOptions SerializerOptions = new()
    {
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
    };

    /// <summary>Reparse a facade rollup list through JSON for a key-order-independent compare.</summary>
    private static string NormalizeList(IReadOnlyList<Rollup> rollups)
    {
        string json = JsonSerializer.Serialize(rollups, SerializerOptions);
        return Normalize(JsonNode.Parse(json)!);
    }

    /// <summary>Canonicalize a JSON node: sort object keys, render numbers uniformly.</summary>
    private static string Normalize(JsonNode node) => Canonical(node).ToJsonString();

    private static JsonNode Canonical(JsonNode node)
    {
        switch (node)
        {
            case JsonObject obj:
                var sorted = new JsonObject();
                foreach (var kv in obj.OrderBy(p => p.Key, StringComparer.Ordinal))
                {
                    sorted[kv.Key] = kv.Value is null ? null : Canonical(kv.Value);
                }
                return sorted;
            case JsonArray arr:
                var outArr = new JsonArray();
                foreach (JsonNode? item in arr)
                {
                    outArr.Add(item is null ? null : Canonical(item));
                }
                return outArr;
            default:
                JsonValue value = node.AsValue();
                if (value.TryGetValue(out double d))
                {
                    return JsonValue.Create(d);
                }
                return JsonValue.Create(value.GetValue<JsonElement>())!;
        }
    }
}
